### PR TITLE
kill global g_windowManager

### PR DESCRIPTION
**Pull request by FernetMenta**
- kills g_windowManager
- created gui components that owns gui objects
- CWindowManager is embedded into CGUIComponent
- StereoscopicManager follows

clients of GUI should check the interface pointer they get from ServiceBroker. GUI won't be mandatory in future version.

### CODE REVIEW

File: xbmc/guilib/GUIComponent.h
Comment (FernetMenta) [inline xbmc/guilib/GUIComponent.h]: @garbear as a gui object, StareoscopicManager should be placed here.

**Comment by garbear** [comment]
I didn't know you'd come to the rescue with a thousand lines of code ;) I'll review this weekend and relocate StereoscopicsManager.

**Review by garbear — APPROVED** [review]
PR looks good. I'll port CStereoscopicsManager in the next day or two

File: xbmc/utils/FileUtils.cpp
Comment (Rechi) [inline xbmc/utils/FileUtils.cpp]: this breaks the test suite

Comment (FernetMenta) [reply xbmc/utils/FileUtils.cpp]: once again the useless test suite. can you point me to the last known case where the test suite was useful by bringing an issue to surface?

Comment (FernetMenta) [reply xbmc/utils/FileUtils.cpp]: but this is a good opportunity to drop a dependency from a utility to gui. I'll fix it

**Comment by FernetMenta** [comment]
@da-anda 014e8a448526784a946bf846aa6fb1bdd34d4126 shows the tedious tasks we have to do. No fance framework or technology solves those issues. GUI logic has to be removed from lower layers.

@garbear more bullet points for your how-to:
- never ever use default parameters in functions/methods. In our case here test suite was not aware that it requested confirmation dialog for DeleteItem
- components like generic utils, network, input, etc, must not depend on GUI.

File: xbmc/interfaces/legacy/ModuleXbmcvfs.cpp
Comment (FernetMenta) [inline xbmc/interfaces/legacy/ModuleXbmcvfs.cpp]: what hack is this? looks like abuse of parameter "force"

**Comment by Voyager1** [comment]
this segfaults on Windows. There are calls to GetWindowManager() from the WM_PAINT event before the CGUIComponent pointer is even initialized.

File: xbmc/Application.cpp
Comment (Voyager1) [inline xbmc/Application.cpp]: to fix the segfault, this should be moved up to the beginning of CApplication::CreateGUI()

Comment (FernetMenta) [reply xbmc/Application.cpp]: windowing has to be initialized before gui. windowing must not depend on gui. if this is the case, it needs to be fixed. as an intermediate sollution it can check if CServiveBroker::GetGUI is nullptr before accessing it.

Comment (Voyager1) [reply xbmc/Application.cpp]: @FernetMenta it may also be sufficient to move the ```m_renderGUI = true;``` statement after the m_pGUI initialisation? I need to check that.

Comment (FernetMenta) [reply xbmc/Application.cpp]: yes, we should move this statement anyway. good catch

Comment (da-anda) [reply xbmc/Application.cpp]: popcornmix disabled the splash screen temporarily for RPI nightlies, which is likely related to this, see https://github.com/popcornmix/xbmc/commit/88d06379af004e33c6c5158b0228c1e78a9c9329

**Comment by DaveTBlake** [comment]
Would it not be better in future to create test builds for all platforms and *do some testing* (or ask others to test for you)  before merging such large changes?

Because of this PR I just spent several hours on a messy rebase of my work, that has been ready for merge for weeks but waiting for review, only to find it seg faults. Very unimpressed.

**Comment by garbear** [comment]
@DaveTBlake killing globals is a thankless task. It takes hours and it's almost guaranteed something will break every time. Nevertheless, it's important that we fight these architectural road blocks. More globals-killing is needed, so keep a lookout for these PRs. Thanks for the understanding.

**Comment by peak3d** [comment]
@garbear yep, I believe that it is an thankless task, and therefore we should produce something for testing. Don't matter if internal or external. I suggest to trigger a test build and ask kodi members to test it on different platforms. If the test version is public available, I could also ask the userbase or we make a thread in forum.kodi.tv for such critical test builds.

**Comment by FernetMenta** [comment]
Well, sorry for the inconvenience but I don't think the trouble was that hard as it seems when reading those complaints. Sorry for the breakage of Windows and PI build and many thanks to @Voyager1 @afedchin @popcornmix for helping out fixing the issues quickly.
Apart from a crash at exit, fix is already provided, current master is fine. Do you really consider this as a major case to worry about? Why not feeling happy that the application has been improved?

**Comment by DaveTBlake** [comment]
>Why not feeling happy that the application has been improved?

Happy for that @FernetMenta, well done have a cookie etc.

Unhappy to have wasted 8 hours of my life, which could have been avoided with a better communication. Unhappy that your response to that is "I don't think the trouble was that hard".  Well that is just your opinion.

Equally that waste could have been avoided with some very basic platform testing before this extensive PR was merged.  Of course testing does not catch everything, but we could at least try. I understand the need for speed to avoid yet more rework, but making test builds available on the mirrors is so easy, and members would use them to test something promptly if asked.

**Pull request merged**
